File: lib/src/test/java/net/mm2d/upnp/SsdpSearchServerTest.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.upnp.SsdpSearchServer.ResponseListener;
import net.mm2d.upnp.SsdpServer.Address;
import net.mm2d.upnp.SsdpServerDelegate.Receiver;
import net.mm2d.util.NetworkUtils;
import net.mm2d.util.TestUtils;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatchers;

import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;

import static org.hamcrest.Matchers.*;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

@RunWith(JUnit4.class)
public class SsdpSearchServerTest {
    @Test
    public void search_ST_ALLでのサーチ() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpServerDelegate delegate = spy(new SsdpServerDelegate(mock(Receiver.class), Address.IP_V4, networkInterface));
        final InterfaceAddress interfaceAddress = TestUtils.createInterfaceAddress("192.0.2.2", "255.255.255.0", 16);
        doReturn(interfaceAddress).when(delegate).getInterfaceAddress();
        final MockMulticastSocket socket = new MockMulticastSocket();
        doReturn(socket).when(delegate).createMulticastSocket(anyInt());
        final SsdpSearchServer server = spy(new SsdpSearchServer(delegate));
        server.open();
        server.start();
        server.search();
        server.stop();
        server.close();

        final DatagramPacket packet = socket.getSendPacket();
        final SsdpRequest message = new SsdpRequest(
                mock(InterfaceAddress.class), packet.getData(), packet.getLength());
        assertThat(message.getMethod(), is(SsdpMessage.M_SEARCH));
        assertThat(message.getHeader(Http.ST), is(SsdpSearchServer.ST_ALL));
        assertThat(message.getHeader(Http.MAN), is(SsdpMessage.SSDP_DISCOVER));
    }

    @Test
    public void search_ST_ROOTDEVICEでのサーチ() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpServerDelegate delegate = spy(new SsdpServerDelegate(mock(Receiver.class), Address.IP_V4, networkInterface));
        final InterfaceAddress interfaceAddress = TestUtils.createInterfaceAddress("192.0.2.2", "255.255.255.0", 16);
        doReturn(interfaceAddress).when(delegate).getInterfaceAddress();
        final MockMulticastSocket socket = new MockMulticastSocket();
        doReturn(socket).when(delegate).createMulticastSocket(anyInt());
        final SsdpSearchServer server = spy(new SsdpSearchServer(delegate));
        server.open();
        server.start();
        server.search(SsdpSearchServer.ST_ROOTDEVICE);
        server.stop();
        server.close();

        final DatagramPacket packet = socket.getSendPacket();
        final SsdpRequest message = new SsdpRequest(
                mock(InterfaceAddress.class), packet.getData(), packet.getLength());
        assertThat(message.getMethod(), is(SsdpMessage.M_SEARCH));
        assertThat(message.getHeader(Http.ST), is(SsdpSearchServer.ST_ROOTDEVICE));
        assertThat(message.getHeader(Http.MAN), is(SsdpMessage.SSDP_DISCOVER));
    }

    @Test
    public void setResponseListener_受信メッセージが通知されること() throws Exception {
        final SsdpServerDelegate delegate = mock(SsdpServerDelegate.class);
        final InterfaceAddress interfaceAddress = TestUtils.createInterfaceAddress("192.0.2.2", "255.255.255.0", 16);
        doReturn(interfaceAddress).when(delegate).getInterfaceAddress();
        final SsdpSearchServer server = spy(new SsdpSearchServer(delegate));
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin");
        final InetAddress address = InetAddress.getByName("192.0.2.2");
        final ArgumentCaptor<SsdpResponse> captor = ArgumentCaptor.forClass(SsdpResponse.class);
        final ResponseListener listener = mock(ResponseListener.class);
        doNothing().when(listener).onReceiveResponse(captor.capture());
        server.setResponseListener(listener);
        server.onReceive(address, data, data.length);

        final SsdpResponse response = captor.getValue();
        assertThat(response.getStatus(), is(Http.Status.HTTP_OK));
        assertThat(response.getUuid(), is("uuid:01234567-89ab-cdef-0123-456789abcdef"));
    }

    @Test
    public void onReceive_listenerがコールされる() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpSearchServer server = new SsdpSearchServer(Address.IP_V4, networkInterface);
        final ResponseListener listener = mock(ResponseListener.class);
        server.setResponseListener(listener);
        final InetAddress address = InetAddress.getByName("192.0.2.2");
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin");

        server.onReceive(address, data, data.length);

        verify(listener, times(1)).onReceiveResponse(ArgumentMatchers.any(SsdpResponse.class));
    }

    @Test
    public void onReceive_アドレス不一致ならlistenerコールされない() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpSearchServer server = new SsdpSearchServer(Address.IP_V4, networkInterface);
        final ResponseListener listener = mock(ResponseListener.class);
        server.setResponseListener(listener);
        final InetAddress address = InetAddress.getByName("192.0.2.3");
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin");

        server.onReceive(address, data, data.length);

        verify(listener, never()).onReceiveResponse(ArgumentMatchers.any(SsdpResponse.class));
    }

    @Test
    public void onReceive_データに問題がある場合listenerコールされない() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpSearchServer server = new SsdpSearchServer(Address.IP_V4, networkInterface);
        final ResponseListener listener = mock(ResponseListener.class);
        server.setResponseListener(listener);
        final InetAddress address = InetAddress.getByName("192.0.2.2");
        final byte[] data = new byte[0];

        server.onReceive(address, data, data.length);

        verify(listener, never()).onReceiveResponse(ArgumentMatchers.any(SsdpResponse.class));
    }

    @Test
    public void onReceive_listenerが登録されていなくてもクラッシュしない() throws Exception {
        final NetworkInterface networkInterface = NetworkUtils.getAvailableInet4Interfaces().get(0);
        final SsdpSearchServer server = new SsdpSearchServer(Address.IP_V4, networkInterface);

        final InetAddress address = InetAddress.getByName("192.0.2.2");
        final byte[] data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin");

        server.onReceive(address, data, data.length);
    }

}

File: mmupnp/src/test/java/net/mm2d/upnp/internal/server/SsdpSearchServerTest.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.server

import com.google.common.truth.Truth.assertThat
import io.mockk.*
import net.mm2d.upnp.SsdpMessage
import net.mm2d.upnp.common.Http
import net.mm2d.upnp.common.util.NetworkUtils
import net.mm2d.upnp.internal.message.SsdpRequest
import net.mm2d.upnp.internal.message.SsdpResponse
import net.mm2d.upnp.internal.thread.TaskExecutors
import net.mm2d.upnp.util.TestUtils
import net.mm2d.upnp.util.createInterfaceAddress
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.net.InetAddress

@Suppress("TestFunctionName", "NonAsciiCharacters")
@RunWith(JUnit4::class)
class SsdpSearchServerTest {
    private lateinit var taskExecutors: TaskExecutors

    @Before
    fun setUp() {
        taskExecutors = TaskExecutors()
    }

    @After
    fun tearDown() {
        taskExecutors.terminate()
    }

    @Test
    fun search_ST_ALLでのサーチ() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val delegate = spyk(SsdpServerDelegate(taskExecutors, Address.IP_V4, networkInterface))
        delegate.setReceiver(mockk(relaxed = true))
        val interfaceAddress = createInterfaceAddress("192.0.2.2", "255.255.255.0", 16)
        every { delegate.interfaceAddress } returns interfaceAddress
        val socket = MockMulticastSocket()
        every { delegate.createMulticastSocket(any()) } returns socket
        val server = spyk(SsdpSearchServer(delegate))
        server.start()
        server.search()
        Thread.sleep(500)
        server.stop()

        val packet = socket.sendPacket!!
        val message = SsdpRequest.create(mockk(relaxed = true), packet.data, packet.length)
        assertThat(message.getMethod()).isEqualTo(SsdpMessage.M_SEARCH)
        assertThat(message.getHeader(Http.ST)).isEqualTo(SsdpSearchServer.ST_ALL)
        assertThat(message.getHeader(Http.MAN)).isEqualTo(SsdpMessage.SSDP_DISCOVER)
    }

    @Test
    fun search_ST_ROOTDEVICEでのサーチ() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val delegate = spyk(SsdpServerDelegate(taskExecutors, Address.IP_V4, networkInterface))
        delegate.setReceiver(mockk(relaxed = true))
        val interfaceAddress = createInterfaceAddress("192.0.2.2", "255.255.255.0", 16)
        every { delegate.interfaceAddress } returns interfaceAddress
        val socket = MockMulticastSocket()
        every { delegate.createMulticastSocket(any()) } returns socket
        val server = spyk(SsdpSearchServer(delegate))
        server.start()
        server.search(SsdpSearchServer.ST_ROOTDEVICE)
        Thread.sleep(500)
        server.stop()

        val packet = socket.sendPacket!!
        val message = SsdpRequest.create(mockk(relaxed = true), packet.data, packet.length)
        assertThat(message.getMethod()).isEqualTo(SsdpMessage.M_SEARCH)
        assertThat(message.getHeader(Http.ST)).isEqualTo(SsdpSearchServer.ST_ROOTDEVICE)
        assertThat(message.getHeader(Http.MAN)).isEqualTo(SsdpMessage.SSDP_DISCOVER)
    }

    @Test
    fun setResponseListener_受信メッセージが通知されること() {
        val delegate: SsdpServerDelegate = mockk(relaxed = true)
        delegate.setReceiver(mockk(relaxed = true))
        val interfaceAddress = createInterfaceAddress("192.0.2.2", "255.255.255.0", 16)
        every { delegate.interfaceAddress } returns interfaceAddress
        every { delegate.getLocalAddress() } returns interfaceAddress.address

        val server = spyk(SsdpSearchServer(delegate))
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin")
        val address = InetAddress.getByName("192.0.2.2")
        val slot = slot<SsdpResponse>()
        val listener: (SsdpMessage) -> Unit = mockk(relaxed = true)
        every { listener.invoke(capture(slot)) } answers { nothing }
        server.setResponseListener(listener)
        server.onReceive(address, data, data.size)

        val response = slot.captured
        assertThat(response.getStatus()).isEqualTo(Http.Status.HTTP_OK)
        assertThat(response.uuid).isEqualTo("uuid:01234567-89ab-cdef-0123-456789abcdef")
    }

    @Test
    fun onReceive_listenerがコールされる() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)
        val listener: (SsdpMessage) -> Unit = spyk()
        server.setResponseListener(listener)
        val address = InetAddress.getByName("192.0.2.2")
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin")

        server.onReceive(address, data, data.size)

        verify(exactly = 1) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_filterでfalseになるとlistenerコールされない() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)
        val listener: (SsdpMessage) -> Unit = spyk()
        server.setResponseListener(listener)
        val address = InetAddress.getByName("192.0.2.2")
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin")
        server.setFilter { false }

        server.onReceive(address, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_アドレス不一致ならlistenerコールされない() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)
        val listener: (SsdpMessage) -> Unit = mockk(relaxed = true)
        server.setResponseListener(listener)
        val address = InetAddress.getByName("192.0.2.3")
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin")

        server.onReceive(address, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_sony_telepathy_serviceは無視する() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)
        val listener: (SsdpMessage) -> Unit = mockk(relaxed = true)
        server.setResponseListener(listener)
        val address = InetAddress.getByName("192.0.2.2")
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response-telepathy.bin")

        server.onReceive(address, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_データに問題がある場合listenerコールされない() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)
        val listener: (SsdpMessage) -> Unit = mockk(relaxed = true)
        server.setResponseListener(listener)
        val address = InetAddress.getByName("192.0.2.2")
        val data = ByteArray(0)

        server.onReceive(address, data, data.size)

        verify(inverse = true) { listener.invoke(any()) }
    }

    @Test
    fun onReceive_listenerが登録されていなくてもクラッシュしない() {
        val networkInterface = NetworkUtils.getAvailableInet4Interfaces()[0]
        val server = SsdpSearchServer(taskExecutors, Address.IP_V4, networkInterface)

        val address = InetAddress.getByName("192.0.2.2")
        val data = TestUtils.getResourceAsByteArray("ssdp-search-response0.bin")

        server.onReceive(address, data, data.size)
    }
}
